### PR TITLE
Add option to RobolectricProcessor to configure if we want to instrument all packages of classes shadowed by a shadow jar by default.

### PR DESCRIPTION
**Pull request by fiower**
Add option to RobolectricProcessor to configure if we want to instrument all packages of classes shadowed by a shadow jar by default.

### CODE REVIEW

**Comment by erd** [comment]
This looks interesting. Could you give me a short summary on why we would want to not instrument the play-services package?

File: robolectric-processor/src/test/java/org/robolectric/annotation/processing/RobolectricProcessorTest.java [hunk 1]
```diff
@@ -32,6 +33,12 @@ public void roboProcessor_supportsPackageOption() {
   }
   
   @Test
+  public void roboProcessor_supportsShouldInstrumentPackageOption() {
+    ASSERT.that(
```
Comment (jongerrish) [inline robolectric-processor/src/test/java/org/robolectric/annotation/processing/RobolectricProcessorTest.java]: org.thruth0?

shouldn't it be org.truth?

Then its assertThat(subject).someAssertion()

Comment (fiower) [inline robolectric-processor/src/test/java/org/robolectric/annotation/processing/RobolectricProcessorTest.java]: I'm following other tests for consistency. Maybe we can update the dependency and import in the future? There should be more usages of org.truth0 in other tests.

**Comment by fiower** [comment]
@erd , we want to be able to test play services components. Instrumenting package will bring unwanted bytecode manipulation.

**Comment by erd** [comment]
How will this affect people that want to be able to create and use play services shadows? I'm fine with adding a flag to the build that enables / disables instrumentation, but I want to make sure that not instrumenting by default is something that will benefit most users.

**Comment by fiower** [comment]
@erd 
People can explicitly add instrumentation rule for play services classes via @Config or properties file, by either instrumenting the entire package(instrumentedPackages) or a single class(shadows). We previously enabled class level instrumentation so for every shadows in the shadows list, its original class is instrumented by default. Play services are optional compared to the android sdk classes. I think it makes sense to let the user decide if a class/package will be instrumented.

**Comment by jongerrish** [comment]
A wider question is, going forward, should we only instrument classes that are shadowed?

The wider net we cast for instrumentation has the side effect of making some final classes non-final, allowing test writers to mock them.... feels like a weird side effect to me.

Personally I'd prefer to tighten this up so that only classes with shadows get instrumented and also to make @Config.instrumentedPackages deprecated / remove it.

Just my option though ;-)

**Comment by erd** [comment]
> The wider net we cast for instrumentation has the side effect of making some final classes non-final, allowing test writers to mock them.... feels like a weird side effect to me.

Absolutely agree!

> Personally I'd prefer to tighten this up so that only classes with shadows get instrumented and also to make @Config.instrumentedPackages deprecated / remove it.

This would be a nice change to investigate.

**Pull request merged**